src/main/java/net/ripe/rpki/rsyncit/rrdp/RrdpFetcher.java

```diff
@@ -226,6 +230,47 @@ private ProcessPublishElementResult processPublishElements(Element doc) throws X
         return new ProcessPublishElementResult(objects, collisionCount.get());
     }
 
-    record ProcessPublishElementResult(List<RpkiObject> objects, int collisionCount) {};
+    record ProcessPublishElementResult(List<RpkiObject> objects, int collisionCount) {}
+
+    public sealed interface FetchResult permits SuccessfulFetch, NoUpdates, FailedFetch {}
+
+    public record SuccessfulFetch(List<RpkiObject> objects, String sessionId, Integer serial) implements FetchResult {}
+    public record NoUpdates(String sessionId, Integer serial) implements FetchResult {}
+    public record FailedFetch(Exception exception) implements FetchResult {}
```

Comment:
Nice to finally have some ADTs in Java :)

Reply:
Yeap, poor man's ADT :)